bitcoin/BRSyncManager.c

```diff
@@ -1820,3 +1883,127 @@ _BRPeerSyncManagerTxPublished (void *info,
                                 { .submitted = {transaction, error} },
                             });
 }
+
+/// MARK: - Misc. Helper Implementations
+
+static BRAddress *
+_getWalletAddresses (BRWallet *wallet,size_t *addressCount) {
+    assert (addressCount);
+
+    size_t addrCount = BRWalletAllAddrs (wallet, NULL, 0);
+
+    BRAddress *addrs = (BRAddress *) calloc (2 * addrCount, sizeof (BRAddress));
+    BRWalletAllAddrs (wallet, addrs, addrCount);
+
+    memcpy (addrs + addrCount, addrs, addrCount * sizeof(BRAddress));
+    for (size_t index = 0; index < addrCount; index++)
+        addrs[addrCount + index] = BRWalletAddressToLegacy (wallet, &addrs[index]);
+
+    *addressCount = 2 * addrCount;
+    return addrs;
+}
+
+static void
+_fillWalletAddressSet(BRSetOf(BRAddress *) addresses, BRWallet *wallet) {
+    size_t addressCount = 0;
+    BRAddress *addressArray = _getWalletAddresses (wallet, &addressCount);
+
+    for (size_t index = 0; index < addressCount; index++) {
+        if (!BRSetContains (addresses, &addressArray[index])) {
+            BRAddress *address = malloc (sizeof(BRAddress));
+            *address = addressArray[index];
+            BRSetAdd (addresses, address);
+        }
+    }
+
+    free (addressArray);
+}
+
+static BRArrayOf(BRAddress *)
+_updateWalletAddressSet(BRSetOf(BRAddress *) addresses, BRWallet *wallet) {
+    size_t addressCount = 0;
+    BRAddress *addressArray = _getWalletAddresses (wallet, &addressCount);
+
+    BRArrayOf(BRAddress *) newAddresses;
+    array_new (newAddresses, addressCount);
+
+    for (size_t index = 0; index < addressCount; index++) {
+        if (!BRSetContains (addresses, &addressArray[index])) {
+            // one copy remains owned by the address set
+            BRAddress *address = malloc (sizeof(BRAddress));
+            *address = addressArray[index];
+            BRSetAdd (addresses, address);
+
+            // one copy owned by the returned array
+            address = malloc (sizeof(BRAddress));
+            *address = addressArray[index];
+            array_add (newAddresses, address);
+        }
+    }
+
+    free (addressArray);
+    return newAddresses;
+}
+
+static uint32_t
+_getLastConfirmedSendTxHeight(BRWallet *wallet,
```

Comment:
This isn't safe because the wallet can arbitrarily remove transactions while were are looping here. I am going to factor this out particular function into the BRWalletManager. Shouldn't affect the rest though.

Reply:
This has been fixed as of commit id ea03c6c. That commit requires the changes in my other PR (https://github.com/breadwallet/breadwallet-core/pull/194) to be put through.